logger.go

```diff
@@ -0,0 +1,234 @@
+	Tee(io.Writer)
+	// Unwrap returns the original proxied target.
+	Unwrap() http.ResponseWriter
+}
+
+// WrapWriter wraps an http.ResponseWriter, returning a proxy that allows you to
+// hook into various parts of the response process.
+func wrapWriter(resp http.ResponseWriter) writerProxy {
+	_, closeNotifier := resp.(http.CloseNotifier)
+	_, flusher := resp.(http.Flusher)
+	_, hiJacker := resp.(http.Hijacker)
+	_, readerFrom := resp.(io.ReaderFrom)
+
+	basicWriter := basicWriter{ResponseWriter: resp}
+	if closeNotifier && flusher && hiJacker && readerFrom {
+		return &fancyWriter{basicWriter}
+	}
+	if flusher {
+		return &flushWriter{basicWriter}
+	}
+	return &basicWriter
+}
+
+// basicWriter wraps a http.ResponseWriter that implements the minimal
+// http.ResponseWriter interface.
+type basicWriter struct {
+	http.ResponseWriter
+	wroteHeader bool
+	code        int
+	bytes       int
+	tee         io.Writer
+}
+
+func (self *basicWriter) WriteHeader(code int) {
+	if !self.wroteHeader {
+		self.code = code
+		self.wroteHeader = true
+		self.ResponseWriter.WriteHeader(code)
+	}
+}
+func (self *basicWriter) Write(buf []byte) (int, error) {
+	self.WriteHeader(http.StatusOK)
+	n, err := self.ResponseWriter.Write(buf)
+	if self.tee != nil {
+		_, err2 := self.tee.Write(buf[:n])
+		// Prefer errors generated by the proxied writer.
+		if err == nil {
+			err = err2
+		}
+	}
+	self.bytes += n
+	return n, err
+}
+func (self *basicWriter) maybeWriteHeader() {
+	if !self.wroteHeader {
+		self.WriteHeader(http.StatusOK)
+	}
+}
+func (self *basicWriter) Status() int {
+	return self.code
+}
+func (self *basicWriter) BytesWritten() int {
+	return self.bytes
+}
+func (self *basicWriter) Tee(writer io.Writer) {
+	self.tee = writer
+}
+func (self *basicWriter) Unwrap() http.ResponseWriter {
+	return self.ResponseWriter
+}
+
+// fancyWriter is a writer that additionally satisfies http.CloseNotifier,
+// http.Flusher, http.Hijacker, and io.ReaderFrom. It exists for the common case
+// of wrapping the http.ResponseWriter that package http gives you, in order to
+// make the proxied object support the full method set of the proxied object.
+type fancyWriter struct {
+	basicWriter
+}
+
+func (self *fancyWriter) CloseNotify() <-chan bool {
+	closeNotifer := self.basicWriter.ResponseWriter.(http.CloseNotifier)
+	return closeNotifer.CloseNotify()
+}
+func (self *fancyWriter) Flush() {
+	flusher := self.basicWriter.ResponseWriter.(http.Flusher)
+	flusher.Flush()
+}
+func (self *fancyWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
+	highJacker := self.basicWriter.ResponseWriter.(http.Hijacker)
+	return highJacker.Hijack()
+}
+func (self *fancyWriter) ReadFrom(reader io.Reader) (int64, error) {
+	if self.basicWriter.tee != nil {
+		return io.Copy(&self.basicWriter, reader)
+	}
+	rf := self.basicWriter.ResponseWriter.(io.ReaderFrom)
+	self.basicWriter.maybeWriteHeader()
+	return rf.ReadFrom(reader)
+}
+
+var _ http.CloseNotifier = &fancyWriter{}
```

Comment:
Can you explain this construct? It's actually for the first time I see "_" in use like this.